File: backend/app/schemas/campaign.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator

from app.db.enums import CampaignStatus


class CampaignCreateRequest(BaseModel):
    name: str = Field(min_length=1)
    template_id: uuid.UUID
    customer_ids: list[uuid.UUID]

    @field_validator("name")
    @classmethod
    def validate_name(cls, value: str) -> str:
        cleaned = value.strip()
        if not cleaned:
            raise ValueError("Campaign name is required.")
        return cleaned


class CampaignRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    template_id: uuid.UUID | None
    template_name: str | None = None
    name: str
    status: CampaignStatus
    recipients_count: int = 0
    customer_ids: list[uuid.UUID] = Field(default_factory=list)
    created_at: datetime
    started_at: datetime | None = None
    finished_at: datetime | None = None
    updated_at: datetime


class CampaignDeleteResult(BaseModel):
    deleted: bool

File: frontend/src/features/campaigns/types/campaign.ts
export type CampaignStatus = 'draft' | 'running' | 'completed' | 'failed'
export type CampaignMessageStatus = 'pending' | 'sent' | 'failed'

export type Campaign = {
  id: string
  user_id: string
  template_id: string | null
  template_name: string | null
  name: string
  status: CampaignStatus
  recipients_count: number
  customer_ids: string[]
  created_at: string
  started_at: string | null
  finished_at: string | null
  updated_at: string
}

export type CampaignsListResponse = {
  data: Campaign[]
}

export type CampaignResponse = {
  data: Campaign
}

export type CampaignCreatePayload = {
  name: string
  template_id: string
  customer_ids: string[]
}

export type CampaignDeleteResponse = {
  data: {
    deleted: boolean
  }
}

export type CampaignSendResponse = {
  data: {
    campaign_id: string
    status: CampaignStatus
    sent: number
    failed: number
  }
}

export type CampaignMessage = {
  id: string
  user_id: string
  campaign_id: string
  customer_id: string | null
  email: string | null
  status: CampaignMessageStatus
  sent_at: string | null
  error_message: string | null
}

export type CampaignMessagesResponse = {
  data: CampaignMessage[]
}
